src/mongo/db/pipeline/expression.cpp

```diff
@@ -490,8 +490,15 @@ const char* ExpressionArray::getOpName() const {
 /* ------------------------- ExpressionArrayElemAt -------------------------- */
 
 Value ExpressionArrayElemAt::evaluate(const Document& root) const {
-    const Value array = vpOperand[0]->evaluate(root);
     const Value indexArg = vpOperand[1]->evaluate(root);
+    long long i = indexArg.coerceToLong();
+
+    // If "vpOperand[0]" is an ExpressionFilter given index is greater than 0 call
+    // filterTillNthValue
+    const Value array = dynamic_cast<ExpressionFilter*>(vpOperand[0].get()) && (i >= 0)
```

Comment:
It's a little strange to do a dynamic cast each and every time you evaluate this. The answer to this won't change, so I'd recommend doing the dynamic cast during optimize, and remembering via a boolean "_firstArgumentIsFilter" or something like that. I'm not sure how slow a dynamic cast is, but certainly not as slow as checking a boolean :)